<!DOCTYPE html> <html lang="en"> <head> <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"> <meta charset="utf-8"> <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"> <meta http-equiv="X-UA-Compatible" content="IE=edge"> <title> De-spaghettification | Christian B. Macdonald, Ph.D. </title> <meta name="author" content="Christian B. Macdonald"> <meta name="description" content="Making pseudo-densities"> <link rel="stylesheet" href="/assets/css/bootstrap.min.css?a4b3f509e79c54a512b890d73235ef04"> <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/css/mdb.min.css" integrity="sha256-jpjYvU3G3N6nrrBwXJoVEYI/0zw8htfFnhT9ljN3JJw=" crossorigin="anonymous"> <link defer rel="stylesheet" href="/assets/css/academicons.min.css?f0b7046b84e425c55f3463ac249818f5"> <link defer rel="stylesheet" type="text/css" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700|Roboto+Slab:100,300,400,500,700|Material+Icons&amp;display=swap"> <link defer rel="stylesheet" href="/assets/css/jekyll-pygments-themes-github.css?591dab5a4e56573bf4ef7fd332894c99" media="" id="highlight_theme_light"> <link rel="stylesheet" href="/assets/css/main.css?d41d8cd98f00b204e9800998ecf8427e"> <link rel="canonical" href="https://odcambc.github.io/blog/2024/fake_densities/"> <script src="/assets/js/theme.js?9a0c749ec5240d9cda97bc72359a72c0"></script> <link defer rel="stylesheet" href="/assets/css/jekyll-pygments-themes-native.css?5847e5ed4a4568527aa6cfab446049ca" media="none" id="highlight_theme_dark"> <script>initTheme();</script> </head> <body class="fixed-top-nav "> <header> <nav id="navbar" class="navbar navbar-light navbar-expand-sm fixed-top" role="navigation"> <div class="container"> <a class="navbar-brand title font-weight-lighter" href="/"> Christian B. Macdonald, Ph.D. </a> <button class="navbar-toggler collapsed ml-auto" type="button" data-toggle="collapse" data-target="#navbarNav" aria-controls="navbarNav" aria-expanded="false" aria-label="Toggle navigation"> <span class="sr-only">Toggle navigation</span> <span class="icon-bar top-bar"></span> <span class="icon-bar middle-bar"></span> <span class="icon-bar bottom-bar"></span> </button> <div class="collapse navbar-collapse text-right" id="navbarNav"> <ul class="navbar-nav ml-auto flex-nowrap"> <li class="nav-item "> <a class="nav-link" href="/">about </a> </li> <li class="nav-item "> <a class="nav-link" href="/blog/">blog </a> </li> <li class="nav-item "> <a class="nav-link" href="/publications/">publications </a> </li> <li class="nav-item "> <a class="nav-link" href="/reviews/">reviews </a> </li> <li class="nav-item "> <a class="nav-link" href="/cv/">cv </a> </li> <li class="nav-item"> <button id="search-toggle" title="Search" onclick="openSearchModal()"> <span class="nav-link">ctrl k <i class="ti ti-search"></i></span> </button> </li> <li class="toggle-container"> <button id="light-toggle" title="Change theme"> <i class="ti ti-sun-moon" id="light-toggle-system"></i> <i class="ti ti-moon-filled" id="light-toggle-dark"></i> <i class="ti ti-sun-filled" id="light-toggle-light"></i> </button> </li> </ul> </div> </div> </nav> <progress id="progress" value="0"> <div class="progress-container"> <span class="progress-bar"></span> </div> </progress> </header> <div class="container mt-5" role="main"> <div class="post"> <header class="post-header"> <h1 class="post-title">De-spaghettification</h1> <p class="post-meta"> Created in June 04, 2024 </p> <p class="post-tags"> <a href="/blog/2024"> <i class="fa-solid fa-calendar fa-sm"></i> 2024 </a> </p> </header> <article class="post-content"> <div id="markdown-content"> <p>Suppose you have, for whatever reason, a bunch of structures that you want to compare. If you have a couple, it’s easy enough to just look at them. What if you have a dozen? Or a thousand? Things become a bit harder. Take this, for example: here, I aligned about 100 various homologous proteins to a reference structure. It’s a bit of a mess.</p> <div class="row mt-3"> <figure> <picture> <source class="responsive-img-srcset" srcset="/assets/img/posts/spaghetti_1-480.webp 480w,/assets/img/posts/spaghetti_1-800.webp 800w,/assets/img/posts/spaghetti_1-1400.webp 1400w," sizes="95vw" type="image/webp"></source> <img src="/assets/img/posts/spaghetti_1.png" class="img-fluid rounded z-depth-1" width="100%" height="auto" loading="eager" onerror="this.onerror=null; $('.responsive-img-srcset').remove();"> </picture> </figure> </div> <div class="caption">Figure 1: a ball of spaghetti.</div> <p>It’s hard to really see what is conserved and what isn’t. To get around this, we can generate psuedo-densities from an aligned set of structures: here, the “density” is essentially (proportional) to a probability density for the overall fold. I think this is sort of cool! It also turns out it’s pretty easy to do. Here’s a short script that will do it for you and a few other instructions.</p> <p>First, align all your structures (duh). Then save them as a single combined PDB file. You can do this easily in PyMOL by choosing “File -&gt; Export Structure -&gt; Export Molecule,” changing the selection to “all” and saving as a PDB file. Make sure you’re writing all objects to one single file on the “Multi-File” tab.</p> <p>Let’s say you saved this as “all_structures.pdb.” You can use the following script to generate a density map from this file in DX format. Install MDAnalysis with <code class="language-plaintext highlighter-rouge">pip install MDAnalysis</code> if you haven’t already. Save this as “generate_density.py” or whatever, and and invoke it (for example) with <code class="language-plaintext highlighter-rouge">python generate_density.py -i all_structures.pdb -o density.dx</code>. You can change the sampling density with the <code class="language-plaintext highlighter-rouge">-d</code> flag.</p> <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kn">import</span> <span class="n">argparse</span>
<span class="kn">import</span> <span class="n">MDAnalysis</span> <span class="k">as</span> <span class="n">mda</span>
<span class="kn">import</span> <span class="n">MDAnalysis.analysis.density</span>

<span class="c1"># read in args from command line using argparse
</span><span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="p">.</span><span class="nc">ArgumentParser</span><span class="p">(</span><span class="n">description</span><span class="o">=</span><span class="sh">'</span><span class="s">This script will take a a pdb file and generate a pseduodensity map in dx format.</span><span class="sh">'</span><span class="p">)</span>
<span class="n">parser</span><span class="p">.</span><span class="nf">add_argument</span><span class="p">(</span><span class="sh">'</span><span class="s">-i</span><span class="sh">'</span><span class="p">,</span> <span class="sh">'</span><span class="s">--input</span><span class="sh">'</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="sh">'</span><span class="s">The input pdb file</span><span class="sh">'</span><span class="p">)</span>
<span class="n">parser</span><span class="p">.</span><span class="nf">add_argument</span><span class="p">(</span><span class="sh">'</span><span class="s">-o</span><span class="sh">'</span><span class="p">,</span> <span class="sh">'</span><span class="s">--output</span><span class="sh">'</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="sh">'</span><span class="s">The name of the output dx file</span><span class="sh">'</span><span class="p">)</span>
<span class="n">parser</span><span class="p">.</span><span class="nf">add_argument</span><span class="p">(</span><span class="sh">'</span><span class="s">-d</span><span class="sh">'</span><span class="p">,</span> <span class="sh">'</span><span class="s">--delta</span><span class="sh">'</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="sh">'</span><span class="s">The grid spacing for sampling PDB density (in Angstroms), default is 0.5</span><span class="sh">'</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
<span class="n">parser</span><span class="p">.</span><span class="nf">add_argument</span><span class="p">(</span><span class="sh">'</span><span class="s">--ca</span><span class="sh">'</span><span class="p">,</span> <span class="sh">'</span><span class="s">--CA</span><span class="sh">'</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="sh">'</span><span class="s">store_true</span><span class="sh">'</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="sh">'</span><span class="s">Use CA (backbone) atoms only</span><span class="sh">'</span><span class="p">)</span>
<span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="p">.</span><span class="nf">parse_args</span><span class="p">()</span>

<span class="n">input_file</span> <span class="o">=</span> <span class="n">args</span><span class="p">.</span><span class="nb">input</span>
<span class="n">output_file</span> <span class="o">=</span> <span class="n">args</span><span class="p">.</span><span class="n">output</span>
<span class="n">delta</span> <span class="o">=</span> <span class="n">args</span><span class="p">.</span><span class="n">delta</span>

<span class="n">pdb_input</span> <span class="o">=</span> <span class="n">mda</span><span class="p">.</span><span class="nc">Universe</span><span class="p">(</span><span class="n">input_file</span><span class="p">)</span>
<span class="k">if</span> <span class="n">args</span><span class="p">.</span><span class="n">ca</span><span class="p">:</span>
    <span class="n">pdb_atoms</span> <span class="o">=</span> <span class="n">pdb_input</span><span class="p">.</span><span class="nf">select_atoms</span><span class="p">(</span><span class="sh">'</span><span class="s">protein and backbone</span><span class="sh">'</span><span class="p">)</span>
<span class="k">else</span><span class="p">:</span>
    <span class="n">pdb_atoms</span> <span class="o">=</span> <span class="n">pdb_input</span><span class="p">.</span><span class="nf">select_atoms</span><span class="p">(</span><span class="sh">'</span><span class="s">protein</span><span class="sh">'</span><span class="p">)</span>

<span class="n">x</span> <span class="o">=</span> <span class="n">MDAnalysis</span><span class="p">.</span><span class="n">analysis</span><span class="p">.</span><span class="n">density</span><span class="p">.</span><span class="nc">DensityAnalysis</span><span class="p">(</span><span class="n">pdb_atoms</span><span class="p">,</span> <span class="n">delta</span><span class="o">=</span><span class="n">delta</span><span class="p">)</span>
<span class="n">x</span><span class="p">.</span><span class="nf">run</span><span class="p">()</span>
<span class="n">x</span><span class="p">.</span><span class="n">density</span><span class="p">.</span><span class="nf">convert_density</span><span class="p">(</span><span class="n">unit</span><span class="o">=</span><span class="sh">'</span><span class="s">Angstrom^{-3}</span><span class="sh">'</span><span class="p">)</span>
<span class="n">x</span><span class="p">.</span><span class="n">density</span><span class="p">.</span><span class="nf">export</span><span class="p">(</span><span class="n">output_file</span><span class="p">)</span>
</code></pre></div></div> <p>Use ChimeraX for the rest. Open the DX file as well as your reference structure. You’ll probably just see a bunch of dots for the map. That’s ok! Go to “Map -&gt; Smooth” and you should see something a little more…mapish. The reference structure will also probably not be aligned with the map. ChimeraX can fit, but it seems to need a little help to start. Move the reference structure until it looks reasonably close, then use “Map -&gt; Fit” to align them. You might need to fiddle but it should work.</p> <p>Play with the gaussian surface levels to get a view of the density. Also try different map styles: image is sort of neat.</p> <div class="row mt-3"> <div class="col-sm mt-3 mt-md-0"> <figure> <picture> <source class="responsive-img-srcset" srcset="/assets/img/posts/backlight_1-480.webp 480w,/assets/img/posts/backlight_1-800.webp 800w,/assets/img/posts/backlight_1-1400.webp 1400w," sizes="95vw" type="image/webp"></source> <img src="/assets/img/posts/backlight_1.png" class="img-fluid rounded z-depth-1" width="100%" height="auto" loading="eager" onerror="this.onerror=null; $('.responsive-img-srcset').remove();"> </picture> </figure> </div> <div class="col-sm mt-3 mt-md-0"> <figure> <picture> <source class="responsive-img-srcset" srcset="/assets/img/posts/backlight_2-480.webp 480w,/assets/img/posts/backlight_2-800.webp 800w,/assets/img/posts/backlight_2-1400.webp 1400w," sizes="95vw" type="image/webp"></source> <img src="/assets/img/posts/backlight_2.png" class="img-fluid rounded z-depth-1" width="100%" height="auto" loading="eager" onerror="this.onerror=null; $('.responsive-img-srcset').remove();"> </picture> </figure> </div> </div> <div class="caption"> Figure 2: Left, pseudo-density for single protein across all solved structures. Right, pseudo-density for protein across 100 homologous proteins, or de-spaghettification of Figure 1. </div> <p>Surface is a bit clearer, if not quite as pretty.</p> <div class="row mt-3"> <div class="col-sm mt-3 mt-md-0"> <figure> <picture> <source class="responsive-img-srcset" srcset="/assets/img/posts/surface_1-480.webp 480w,/assets/img/posts/surface_1-800.webp 800w,/assets/img/posts/surface_1-1400.webp 1400w," sizes="95vw" type="image/webp"></source> <img src="/assets/img/posts/surface_1.png" class="img-fluid rounded z-depth-1" width="100%" height="auto" loading="eager" onerror="this.onerror=null; $('.responsive-img-srcset').remove();"> </picture> </figure> </div> </div> <div class="caption"> Figure 3: Surface view of Figure 2 right, pseudo-density of Figure 1. </div> </div> </article> </div> </div> <footer class="fixed-bottom" role="contentinfo"> <div class="container mt-0"> © Copyright 2024 Christian B. Macdonald. Powered by <a href="https://jekyllrb.com/" target="_blank" rel="external nofollow noopener">Jekyll</a> with <a href="https://github.com/alshedivat/al-folio" rel="external nofollow noopener" target="_blank">al-folio</a> theme. Hosted by <a href="https://pages.github.com/" target="_blank" rel="external nofollow noopener">GitHub Pages</a>. Last updated: September 12, 2024. </div> </footer> <script src="https://cdn.jsdelivr.net/npm/jquery@3.6.0/dist/jquery.min.js" integrity="sha256-/xUj+3OJU5yExlq6GSYGSHk7tPXikynS7ogEvDej/m4=" crossorigin="anonymous"></script> <script src="/assets/js/bootstrap.bundle.min.js"></script> <script src="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/js/mdb.min.js" integrity="sha256-NdbiivsvWt7VYCt6hYNT3h/th9vSTL4EDWeGs5SN3DA=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/masonry-layout@4.2.2/dist/masonry.pkgd.min.js" integrity="sha256-Nn1q/fx0H7SNLZMQ5Hw5JLaTRZp0yILA/FRexe19VdI=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/imagesloaded@5.0.0/imagesloaded.pkgd.min.js" integrity="sha256-htrLFfZJ6v5udOG+3kNLINIKh2gvoKqwEhHYfTTMICc=" crossorigin="anonymous"></script> <script defer src="/assets/js/masonry.js" type="text/javascript"></script> <script defer src="https://cdn.jsdelivr.net/npm/medium-zoom@1.1.0/dist/medium-zoom.min.js" integrity="sha256-ZgMyDAIYDYGxbcpJcfUnYwNevG/xi9OHKaR/8GK+jWc=" crossorigin="anonymous"></script> <script defer src="/assets/js/zoom.js?85ddb88934d28b74e78031fd54cf8308"></script> <script src="/assets/js/no_defer.js?2781658a0a2b13ed609542042a859126"></script> <script defer src="/assets/js/common.js?b7816bd189846d29eded8745f9c4cf77"></script> <script defer src="/assets/js/copy_code.js?12775fdf7f95e901d7119054556e495f" type="text/javascript"></script> <script defer src="/assets/js/jupyter_new_tab.js?d9f17b6adc2311cbabd747f4538bb15f"></script> <script async src="https://d1bxh8uas1mnw7.cloudfront.net/assets/embed.js"></script> <script async src="https://badge.dimensions.ai/badge.js"></script> <script type="text/javascript">window.MathJax={tex:{tags:"ams"}};</script> <script defer type="text/javascript" id="MathJax-script" src="https://cdn.jsdelivr.net/npm/mathjax@3.2.0/es5/tex-mml-chtml.min.js" integrity="sha256-rjmgmaB99riUNcdlrDtcAiwtLIojSxNyUFdl+Qh+rB4=" crossorigin="anonymous"></script> <script defer src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6" crossorigin="anonymous"></script> <script type="text/javascript">function progressBarSetup(){"max"in document.createElement("progress")?(initializeProgressElement(),$(document).on("scroll",function(){progressBar.attr({value:getCurrentScrollPosition()})}),$(window).on("resize",initializeProgressElement)):(resizeProgressBar(),$(document).on("scroll",resizeProgressBar),$(window).on("resize",resizeProgressBar))}function getCurrentScrollPosition(){return $(window).scrollTop()}function initializeProgressElement(){let e=$("#navbar").outerHeight(!0);$("body").css({"padding-top":e}),$("progress-container").css({"padding-top":e}),progressBar.css({top:e}),progressBar.attr({max:getDistanceToScroll(),value:getCurrentScrollPosition()})}function getDistanceToScroll(){return $(document).height()-$(window).height()}function resizeProgressBar(){progressBar.css({width:getWidthPercentage()+"%"})}function getWidthPercentage(){return getCurrentScrollPosition()/getDistanceToScroll()*100}const progressBar=$("#progress");window.onload=function(){setTimeout(progressBarSetup,50)};</script> <script src="/assets/js/tabs.min.js?b8748955e1076bbe0dabcf28f2549fdc"></script> <script src="/assets/js/vanilla-back-to-top.min.js?f40d453793ff4f64e238e420181a1d17"></script> <script> </script> <script type="module" src="/assets/js/search/ninja-keys.min.js?601a2d3465e2a52bec38b600518d5f70"></script> <ninja-keys hidebreadcrumbs noautoloadmdicons placeholder="Type to start searching"></ninja-keys> <script>let theme=determineComputedTheme();const ninjaKeys=document.querySelector("ninja-keys");"dark"===theme?ninjaKeys.classList.add("dark"):ninjaKeys.classList.remove("dark");const openSearchModal=()=>{const e=document.querySelector(".navbar-collapse");e.classList.contains("show")&&e.classList.remove("show"),ninjaKeys.open()};</script> <script>const ninja=document.querySelector("ninja-keys");ninja.data=[{id:"nav-about",title:"about",section:"Navigation",handler:()=>{window.location.href="/"}},{id:"nav-blog",title:"blog",description:"",section:"Navigation",handler:()=>{window.location.href="/blog/"}},{id:"nav-publications",title:"publications",description:"a list of current publications.",section:"Navigation",handler:()=>{window.location.href="/publications/"}},{id:"nav-reviews",title:"reviews",description:"public peer reviews.",section:"Navigation",handler:()=>{window.location.href="/reviews/"}},{id:"nav-cv",title:"cv",description:"Click the button to download my CV. References and contact information are available upon request.",section:"Navigation",handler:()=>{window.location.href="/cv/"}},{id:"post-de-spaghettification",title:"De-spaghettification",description:"Making pseudo-densities",section:"Posts",handler:()=>{window.location.href="/blog/2024/fake_densities/"}},{id:"post-pymol-mapping-notes",title:"PyMOL mapping notes",description:"Reminders for pymol b factor spectrum mapping",section:"Posts",handler:()=>{window.location.href="/blog/2024/more_pymol-copy/"}},{id:"post-note-to-self",title:"Note to self",description:"I learned something new about ls",section:"Posts",handler:()=>{window.location.href="/blog/2024/note-to-self/"}},{id:"post-some-one-liners",title:"Some one-liners",description:"Some fastq/fasta manipulations",section:"Posts",handler:()=>{window.location.href="/blog/2024/commands/"}},{id:"post-sanger-to-fastq",title:"Sanger to fastq",description:"A quick script to convert ab1 files to fastq",section:"Posts",handler:()=>{window.location.href="/blog/2024/abi/"}},{id:"post-some-useful-pymol-defaults",title:"Some useful PyMOL defaults",description:"dot pymolrc",section:"Posts",handler:()=>{window.location.href="/blog/2024/pymol/"}},{id:"post-oct1-paper-is-out",title:"OCT1 paper is out",description:"OCT1 paper is out",section:"Posts",handler:()=>{window.location.href="/blog/2024/oct1/"}},{id:"project-project-1",title:"project 1",description:"with background image",section:"Projects",handler:()=>{window.location.href="/projects/1_project/"}},{id:"project-project-2",title:"project 2",description:"a project with a background image and giscus comments",section:"Projects",handler:()=>{window.location.href="/projects/2_project/"}},{id:"project-project-3-with-very-long-name",title:"project 3 with very long name",description:"a project that redirects to another website",section:"Projects",handler:()=>{window.location.href="/projects/3_project/"}},{id:"project-project-4",title:"project 4",description:"another without an image",section:"Projects",handler:()=>{window.location.href="/projects/4_project/"}},{id:"project-project-5",title:"project 5",description:"a project with a background image",section:"Projects",handler:()=>{window.location.href="/projects/5_project/"}},{id:"project-project-6",title:"project 6",description:"a project with no image",section:"Projects",handler:()=>{window.location.href="/projects/6_project/"}},{id:"socials-email",title:"Send email",section:"Socials",handler:()=>{window.open("mailto:%63%68%72%69%73%74%69%61%6E.%6D%61%63%64%6F%6E%61%6C%64 %61%74 %75%63%73%66.%65%64%75","_blank")}},{id:"socials-orcid",title:"ORCID",section:"Socials",handler:()=>{window.open("https://orcid.org/0000-0002-0201-8832","_blank")}},{id:"socials-google-scholar",title:"Google Scholar",section:"Socials",handler:()=>{window.open("https://scholar.google.com/citations?user=pyTBwS0AAAAJ","_blank")}},{id:"socials-github",title:"GitHub",section:"Socials",handler:()=>{window.open("https://github.com/odcambc","_blank")}},{id:"socials-rss",title:"RSS Feed",section:"Socials",handler:()=>{window.open("/feed.xml","_blank")}},{id:"light-theme",title:"Change theme to light",description:"Change the theme of the site to Light",section:"Theme",handler:()=>{setThemeSetting("light")}},{id:"dark-theme",title:"Change theme to dark",description:"Change the theme of the site to Dark",section:"Theme",handler:()=>{setThemeSetting("dark")}},{id:"system-theme",title:"Use system default theme",description:"Change the theme of the site to System Default",section:"Theme",handler:()=>{setThemeSetting("system")}}];</script> </body> </html>